bson/primitive/primitive.go

```diff
@@ -208,6 +209,55 @@ func (d D) Map() M {
 	return m
 }
 
+// MarshalJSON encodes D into JSON.
+func (d D) MarshalJSON() ([]byte, error) {
+	if d == nil {
+		return json.Marshal(nil)
+	}
+	var err error
+	var buf bytes.Buffer
+	buf.Write([]byte("{"))
+	enc := json.NewEncoder(&buf)
+	for i, e := range d {
+		err = enc.Encode(e.Key)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write([]byte(":"))
+		err = enc.Encode(e.Value)
+		if err != nil {
+			return nil, err
+		}
+		if i < len(d)-1 {
+			buf.Write([]byte(","))
+		}
+	}
+	buf.Write([]byte("}"))
+	return json.RawMessage(buf.Bytes()).MarshalJSON()
+}
+
+// UnmarshalJSON decodes D from JSON.
+func (d *D) UnmarshalJSON(b []byte) error {
+	dec := json.NewDecoder(bytes.NewReader(b))
+	t, err := dec.Token()
+	if err != nil {
+		return err
+	}
+	if t == nil {
+		*d = nil
+		return nil
+	}
+	if v, ok := t.(json.Delim); !ok || v != '{' {
+		return &json.UnmarshalTypeError{
+			Value:  tokenString(t),
+			Type:   reflect.TypeOf(*d),
+			Offset: dec.InputOffset(),
+		}
+	}
+	*d, err = jsonDecodeD(dec)
+	return err
```

Comment:
[Optional] You could combine the logic in D.UnmarshalJSON and m.UnmarshalJSON using generics:

```go
func newJsonObjDecoder[T D | M](b []byte, val *T) (*json.Decoder, error) {
	dec := json.NewDecoder(bytes.NewReader(b))

	t, err := dec.Token()
	if err != nil {
		return nil, err
	}

	if t == nil {
		*val = nil
		return nil, nil
	}

	if v, ok := t.(json.Delim); !ok || v != '{' {
		return nil, &json.UnmarshalTypeError{
			Value:  tokenString(t),
			Type:   reflect.TypeOf(*val),
			Offset: dec.InputOffset(),
		}
	}

	return dec, nil
}
```